src/app/components/jsx-helpers/context-loader.js

```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+import {LoaderPage} from '~/components/jsx-helpers/jsx-helpers.jsx';
+
+/*
+    Load slug
+    Create context value from loaded data
+    Render children within context provider
+*/
+
+function ContextProvider({Context, data, useContextValue, children}) {
+    return (
+        <Context.Provider value={useContextValue(data)}>
+            {children}
+        </Context.Provider>
+    );
+}
+
+export default function ContextLoader({Context, slug, useContextValue, children, ...loaderArgs}) {
+    const Child = ({data}) => (
+        <ContextProvider {...{Context, data, useContextValue, children}} />
+    );
+
+    return (
+        <LoaderPage slug={slug} Child={Child} {...loaderArgs} />
+    );
+}
```

Comment:
I just wrote a little loader for labs that you might find interesting.  It has a built-in delay so that it only displays after a given timeout.  That's helpful for preventing the case where the API returns very quickly and you get a flash of a loader before the content snaps in.  

https://github.com/openstax/labs/blob/main/frontend/src/components/loading-animation.tsx#L185-L195 and https://github.com/openstax/labs/blob/main/frontend/src/lib/hooks.ts#L3